tests/unit/test_kubernetes_control_plane.py

```diff
@@ -76,31 +95,45 @@ def _svc(clusterIP):
 
 
 def test_service_cidr_greenfield_deploy():
-    configure_apiserver(None, "10.152.183.0/24")
+    cni = configure_apiserver(None, "10.152.183.0/24")
+    cni.set_service_cidr.assert_called_once_with("10.152.183.0/24")
     assert not is_flag_set("kubernetes-control-plane.had-service-cidr-expanded")
-    configure_apiserver(None, "10.152.183.0/24,fe80::/120")
+    cni = configure_apiserver(None, "10.152.183.0/24,fe80::/120")
+    cni.set_service_cidr.assert_called_once_with("10.152.183.0/24,fe80::/120")
     assert not is_flag_set("kubernetes-control-plane.had-service-cidr-expanded")
 
 
 def test_service_cidr_no_change():
-    configure_apiserver("10.152.183.0/24", "10.152.183.0/24")
+    cni = configure_apiserver("10.152.183.0/24", "10.152.183.0/24")
+    cni.set_service_cidr.assert_called_once_with("10.152.183.0/24")
     assert not is_flag_set("kubernetes-control-plane.had-service-cidr-expanded")
-    configure_apiserver("10.152.183.0/24,fe80::/120", "10.152.183.0/24,fe80::/120")
+    cni = configure_apiserver(
+        "10.152.183.0/24,fe80::/120", "10.152.183.0/24,fe80::/120"
+    )
+    cni.set_service_cidr.assert_called_once_with("10.152.183.0/24,fe80::/120")
     assert not is_flag_set("kubernetes-control-plane.had-service-cidr-expanded")
 
 
 def test_service_cidr_non_expansion():
-    configure_apiserver("10.152.183.0/24", "10.154.183.0/24")
+    cni = configure_apiserver("10.152.183.0/24", "10.154.183.0/24")
+    cni.set_service_cidr.assert_called_once_with("10.152.183.0/24")
     assert not is_flag_set("kubernetes-control-plane.had-service-cidr-expanded")
-    configure_apiserver("10.152.183.0/24,fe80::/120", "10.152.183.0/24,fe81::/120")
+    cni = configure_apiserver(
+        "10.152.183.0/24,fe80::/120", "10.152.183.0/24,fe81::/120"
+    )
+    cni.set_service_cidr.assert_called_once_with("10.152.183.0/24,fe80::/120")
     assert not is_flag_set("kubernetes-control-plane.had-service-cidr-expanded")
 
 
 def test_service_cidr_expansion():
-    configure_apiserver("10.152.183.0/24", "10.152.0.0/16")
+    cni = configure_apiserver("10.152.183.0/24", "10.152.0.0/16")
+    cni.set_service_cidr.assert_called_once_with("10.152.0.0/16")
     assert is_flag_set("kubernetes-control-plane.had-service-cidr-expanded")
     clear_flag("kubernetes-control-plane.had-service-cidr-expanded")
-    configure_apiserver("10.152.183.0/24,fe80::/120", "10.152.183.0/24,fe80::/112")
+    cni = configure_apiserver(
+        "10.152.183.0/24,fe80::/120", "10.152.183.0/24,fe80::/112"
+    )
+    cni.set_service_cidr.assert_called_once_with("10.152.183.0/24,fe80::/112")
```

Comment:
@Cynerva 
I'm not sure if these two are sharing the correct `service-cidr`.  Can you confirm what the difference between the db value and config value and why one is picked over the other?

Reply:
Most of the time, the two values are the same. However, when the charm starts kube-apiserver for the first time, the charm "freezes" the service CIDR (by setting the DB value) and no longer allows users to change it. The code around that is [here](https://github.com/charmed-kubernetes/charm-kubernetes-control-plane/blob/fb4460092b20e1151ee30672f3bdd3e4366717ed/lib/charms/layer/kubernetes_control_plane.py#L405-L416).

We do this because changing the service CIDR in an existing cluster is extremely disruptive. It breaks services in Kubernetes. There is one exception: if the service CIDR change is an "expansion" (the new CIDR is a superset of the old CIDR) then the change is safe and the charm will allow it.

The assertions in these tests look correct to me. We're sharing the actual value that gets passed to kube-apiserver. That's what we need kube-ovn to match up with.

Reply:
AH!!!  yes that explains what's going on.  Cool!  Thanks for the 👀